crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -581,7 +581,11 @@ impl<'w> EntityMut<'w> {
     ///
     /// # Safety
     ///
-    /// - `T` must be a mutable component
+    /// - One of the following should be true:
+    ///   - Either `T` is mutable, or
+    ///   - `OnReplace` hooks and observers for that component on that entity should trigger immediately before the mutation
+    ///     and `OnInsert` should trigger immediately after the mutation, or
+    ///   - The user should uphold documented invariants of `T`. If no such documentation provided, it is impossible for the end user to uphold this.
```

Comment:
Is this requirement necessary?  If it is, then I think `fn modify_component()` would need to be `unsafe` and include a similar requirement.  

But it seems like this could be handled by privacy: Anything that would violate the invariants of `T` just shouldn't be `pub`.  That just leaves things like `mem::swap`, but as long as you're firing triggers that should be indistinguishable from overwriting the component with `insert`.